src/os/bluestore/KernelDevice.cc

```diff
@@ -812,18 +812,39 @@ int KernelDevice::_sync_write(uint64_t off, bufferlist &bl, bool buffered, int w
   }
   vector<iovec> iov;
   bl.prepare_iov(&iov);
-  int r = ::pwritev(choose_fd(buffered, write_hint),
-		    &iov[0], iov.size(), off);
 
-  if (r < 0) {
-    r = -errno;
-    derr << __func__ << " pwritev error: " << cpp_strerror(r) << dendl;
-    return r;
-  }
+  auto left = len;
+  auto o = off;
+  size_t pos = 0;
+  do {
+    auto r = ::pwritev(choose_fd(buffered, write_hint),
+      &iov[pos], iov.size() - pos, o);
+
+    if (r < 0) {
+      r = -errno;
+      derr << __func__ << " pwritev error: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+    o += r;
+    left -= r;
+    if (left) {
+      while (r) {
+        ceph_assert(pos < iov.size());
+        if ((size_t)r < iov[pos].iov_len) {
```

Comment:
NIT: I think that loop should skip processed iovs, and action outside loop adapt subsequent one iov.

Reply:
If pwrite was not fully successful, we do following:
1) skip fully processed iovs
2) update at most one iov that was cut 
3) repeat pwrite with remaining iovs
I see that 1) and 2) are done in common loop. 
Proposal was (of course if you think that it would be clearer) to split this into steps:
1) loop that skips fully processed iovs
2) change that one iov

Reply:
OK, thanks for the suggestion. Done